src/Client.cpp

```diff
@@ -93,120 +71,105 @@ void UrsaClient::init_conn(zmq::socket_t &socket) {
     }
 }
 
-void UrsaClient::recv_res(zmq::socket_t &socket) {
-    // enable periodic progress check
-    // which is done by status worker
-    this->command_active = true;
-    this->wait_time = 0;
-
-    while (this->command_active) {
-        auto res_str = s_recv<std::string>(&socket);
-
-        if (res_str.empty()) {
-            continue;
+void UrsaClient::recv_res(zmq::socket_t *socket) {
+    std::string result;
+    while (result.empty()) {
+        try {
+            result = s_recv<std::string>(socket);
+        } catch (const std::runtime_error &ex) {
+            check_task_status(this->connection_id);
         }
+    }
 
-        this->command_active = false;
-        auto res = json::parse(res_str);
-
-        if (this->raw_json) {
-            std::cout << res.dump(4) << std::endl;
-            return;
-        }
+    auto res = json::parse(result);
+    if (this->raw_json) {
+        std::cout << res.dump(4) << std::endl;
+        return;
+    }
 
-        if (res["type"] == "select") {
-            for (const auto &file : res["result"]["files"]) {
-                std::cout << file.get<std::string>() << std::endl;
-            }
-        } else if (res["type"] == "error") {
-            spdlog::error(res["error"]["message"].get<std::string>());
-        } else {
-            std::cout << res.dump(4) << std::endl;
+    if (res["type"] == "select") {
```

Comment:
I know you didn't chnage it, but there is a problem when this is a response to `select into iterator`.

Maybe:
```suggestion
    if (res["type"] == "select" && !res.value("iterator", false)) {
```

```suggestion
    if (res["type"] == "select" && res.find("iterator") == res.end()) {
```


(didn't compile them myself, but should work)